### PR TITLE
Optimized polyak updates

### LINKED ISSUE #93: [enhancement] Polyak Averaging could be done faster
This is rather minor, but polyak averaging in DQN/SAC/TD3 could be done faster with far fewer intermediate tensors using `torch.addcmul_` https://pytorch.org/docs/stable/torch.html#torch.addcmul.

### PR DESCRIPTION
**Pull request by PartiallyTyped**
Optimized Polyak updates through in-place operations.

## Description
Used `th.no_grad` to avoid storing the graph. Used in-place `th.Tensor.mul` and `th.add` operations to avoid creating extra tensors.

Achieves a speedup of 1.5 and 1.95 for cpus and gpus in google colab over the old version.

## Motivation and Context

- [x] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

Closes #93 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [x] Optimization

## Checklist:

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [x] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [x] I have checked the codestyle using `make lint`
- [] I have ensured `make pytest` and `make type` both pass.

### CODE REVIEW

File: tests/test_utils.py [hunk 1]
```diff
@@ -143,3 +145,16 @@ def test_vec_noise():
         vec.noises = [base] * (num_envs - 1)
     assert all(isinstance(noise, type(base)) for noise in vec.noises)
     assert len(vec.noises) == num_envs
+
+
+def test_polyak():
+    p1, p2 = th.nn.Parameter(th.ones((5, 5))), th.nn.Parameter(th.zeros((5, 5)))
```
Comment (araffin) [inline tests/test_utils.py:151]: please use `param_1` and `target_1`

File: stable_baselines3/common/utils.py [hunk 2]
```diff
@@ -271,3 +271,20 @@ def safe_mean(arr: Union[np.ndarray, list, deque]) -> np.ndarray:
...
+    """
+    Perform a polyak average update on target_params using params
+
+    :param params: (Iterable[th.nn.Parameter]) parameters to use to update the target params
+    :param target_params: (Iterable[th.nn.Parameter]) parameters to update
+    :param tau: (float) the soft update coefficient ("Polyak update", between 0 and 1)
+    """
+    params = list(params)
+    target_params = list(target_params)
+    if params:
```
Comment (araffin) [inline stable_baselines3/common/utils.py:286]: maybe an assert instead? "assert len(params) > 0" + meaningful error message?

File: stable_baselines3/common/utils.py [hunk 3]
```diff
@@ -271,3 +271,20 @@ def safe_mean(arr: Union[np.ndarray, list, deque]) -> np.ndarray:
     :return:
     """
     return np.nan if len(arr) == 0 else np.mean(arr)
+
+
+def polyak_update(params: Iterable[th.nn.Parameter], target_params: Iterable[th.nn.Parameter], tau: float):
```
Comment (araffin) [inline stable_baselines3/common/utils.py:276]:
```suggestion
def polyak_update(params: Iterable[th.nn.Parameter], target_params: Iterable[th.nn.Parameter], tau: float) -> None:
```

File: stable_baselines3/common/utils.py (hunk 2 above)
Comment (PartiallyTyped) [reply stable_baselines3/common/utils.py:286]: Not sure if it's necessary, the if is simply to get the device, but besides that, it isn't necessary.

Comment (PartiallyTyped) [reply stable_baselines3/common/utils.py:286]: Never mind my previous reply, the if was garbage left from a different implementation.

Comment (PartiallyTyped) [reply stable_baselines3/common/utils.py:286]: The previous implementation used the first parameter to get the device and create a tensor with `tau`, but this version doesn't need it. I forgot about it and accidentally left it in.

File: stable_baselines3/common/utils.py [hunk 4]
```diff
@@ -284,3 +284,17 @@ def safe_mean(arr: Union[np.ndarray, list, deque]) -> np.ndarray:
     :return:
     """
     return np.nan if len(arr) == 0 else np.mean(arr)
+
+
+def polyak_update(params: Iterable[th.nn.Parameter], target_params: Iterable[th.nn.Parameter], tau: float) -> None:
+    """
+    Perform a polyak average update on target_params using params
```
Comment (araffin) [inline stable_baselines3/common/utils.py:291]:
```suggestion
    Perform a Polyak average update on ``target_params`` using ``params``
```

+ maybe add a comment of why we are doing it that way?
+ maybe a link to the issue?

Comment (PartiallyTyped) [reply stable_baselines3/common/utils.py:291]: Will add both!

File: stable_baselines3/common/utils.py [hunk 5]
```diff
@@ -284,3 +284,17 @@ def safe_mean(arr: Union[np.ndarray, list, deque]) -> np.ndarray:
...
+
```
Comment (PartiallyTyped) [inline stable_baselines3/common/utils.py:292]:
```suggestion
    The Polyak update is done in place, with ``no_grad``, and therefore does not create intermediate tensors, or a computation graph, reducing memory cost and improving performance.  We scale the target params by ``1-tau`` (in-place), add the new weights, scaled by ``tau`` and store the result of the sum in the target params (in place). 
```

File: stable_baselines3/common/utils.py [hunk 6]
```diff
@@ -284,3 +284,20 @@ def safe_mean(arr: Union[np.ndarray, list, deque]) -> np.ndarray:
...
     return np.nan if len(arr) == 0 else np.mean(arr)
+
+
+def polyak_update(params: Iterable[th.nn.Parameter], target_params: Iterable[th.nn.Parameter], tau: float) -> None:
+    """
+    Perform a Polyak average update on ``target_params`` using ``params``
+    The Polyak update is done in place, with ``no_grad``, and therefore does not create intermediate tensors, 
+    or a computation graph, reducing memory cost and improving performance.  We scale the target params 
+    by ``1-tau`` (in-place), add the new weights, scaled by ``tau`` and store the result of the sum in the target 
+    params (in place).
```
Comment (PartiallyTyped) [inline stable_baselines3/common/utils.py:295]:
```suggestion
    params (in place). 
    See https://github.com/DLR-RM/stable-baselines3/issues/93
```

**Comment by araffin** [comment]
@PartiallyTyped i think we are done with that one, no?

**Comment by PartiallyTyped** [comment]
LGTM :)

**Comment by araffin** [comment]
I will just do a quick performance check (one seed on two bullet env with SAC) and then merge ;)

**Pull request merged**